### ISSUE TITLE
Does not accept import() in function scope

### ISSUE THREAD

**Issue by diachedelic** [douglascrockford/JSLint#270]
```js
/*jslint node */
(function () {
    return import("./foo.js").then(console.log);
}());
```
fails JSLint with
> Undeclared 'import'.

There does not appear to be a workaround.

**Comment by douglascrockford** [comment]
JSLint expects all imports at the top level.

**Comment by diachedelic** [comment]
I moved the import to the top level, but it still fails.

```
/*jslint node */
const transformerImport = import(process.argv[2]);

function loadTransformer() {
    return function loadTransformerRequestor(callback) {
        return transformerImport.then(
            (module) => callback(module.default)
        ).catch(
            (err) => callback(undefined, err)
        );
    };
}

// some complex eventual control flow here, which uses loadTransformer
```

> Undeclared 'import'.

**Comment by kaizhu256** [comment]
tweaking the jslint-headers should fix your problem:

```diff
-/*jslint node */
+/*jslint browser, node */
+/*global import*/
 const transformerImport = import(process.argv[2]);
 
 function loadTransformer() {
     return function loadTransformerRequestor(callback) {
         return transformerImport.then(
             (module) => callback(module.default)
         ).catch(
             (err) => callback(undefined, err)
         );
     };
 }
```

![image](https://user-images.githubusercontent.com/280571/92532456-efc26f00-f1f5-11ea-96db-7fe17ecac7fb.png)

**Comment by diachedelic** [comment]
Ah but I am using modules.

```
/*jslint browser, node */
/*global import*/
import foo from "foo";
const transformerImport = import(process.argv[2]);

function loadTransformer() {
    return function loadTransformerRequestor(callback) {
        return transformerImport.then(
            (module) => callback(module.default)
        ).catch(
            (err) => callback(undefined, err)
        );
    };
}
```

> When using modules, don't use directive '/*global'.

**Comment by kaizhu256** [comment]
i sympathize with your use-case.  i found 2 solutions, but neither of them is really great:

1. is it possible to use `require()` instead of `import()`?
```diff
-const transformerImport = import(process.argv[2]);
+const transformerImport = require(process.argv[2]);
```

2. else hack jslint's global-list to include import:
```diff
--- a/jslint.js
+++ b/jslint.js

     bitwise: true,
     browser: [
+        "import",
         "caches", "CharacterData", "clearInterval", "clearTimeout", "document",
         "DocumentType", "DOMException", "Element", "Event", "event", "fetch",
         "FileReader", "FontFace", "FormData", "history", "IntersectionObserver",
...
     getset: true,
     long: true,
     node: [
+        "import",
         "Buffer", "clearImmediate", "clearInterval", "clearTimeout",
         "console", "exports", "module", "process", "require",
         "setImmediate", "setInterval", "setTimeout", "TextDecoder",
```

**Comment by diachedelic** [comment]
> is it possible to use require() instead of import()?
Unfortunately not - `require()` cannot be used to import ES modules.
> else hack jslint's global-list to include import:
Thanks - perhaps a temporary workaround.

I would like an explanation of why the functionality of `import()` is so restricted - if it can only be used in the top level, it becomes practically useless as far as I can tell. The whole point of `import()` is that you can fetch modules which you only find out the path of at runtime.

However, I do think it's important to prevent misuse of `import()` as `require()` is so commonly misued - e.g. `require("fs").readFile(...)`.

**Comment by diachedelic** [comment]
Here is my program in full, I'm sorry I did not include it initially. Please tell me how I can avoid using function-level `import()`, or explain why I do not need to use it. I really want to pass JSLint! Thanks  @douglascrockford .

```js
/*jslint node */
import parseq from "./parseq.js";
import make_capabilities from "./capabilities.js";

// Runs a transform over all files tracked by git in the current working
// directory.
// Usage: $0 <transformer_name>
// where <transformer_name> selects to a module exporting a requestor factory
// like the example below.
//
// Contrived example: uppercase the source code of all JS files, and add a
// ".upper" file extension.
//
// function upper_case_transformer() {
//     return function upper_case_transformer_requestor(
//         callback,
//         {file_path, read_source_requestor, write_source_requestor}
//     ) {
//         if (!file_path.endsWith(".js")) {
//             return callback(file_path);
//         }

//         return parseq.sequence([
//             read_source_requestor,
//             function transform_source_requestor(callback, source) {
//                 return callback(source.toUpperCase());
//             },
//             write_source_requestor,
//             function transform_filename_requestor(callback) {
//                 return callback(file_path + ".upper");
//             }
//         ])(callback, file_path);
//     };
// }

function load_transformer(name) {
    return function load_transformer_requestor(callback) {
        return import(
            "./transformers/" + name.replace("/", "") + ".js"
        ).then(
            (module) => callback(module.default)
        ).catch(
            (err) => callback(undefined, err)
        );
    };
}

const {
    clean,
    delete_empty_directories,
    list_tracked_files,
    read_source,
    write_source,
    rename_file,
    stage_all
} = make_capabilities(process.cwd());

function callback(value, reason) {
    if (value === undefined) {
        console.error(reason);
        return process.exit(1);
    }

    return process.exit(0);
}

if (process.argv[2] === "--clean") {
    clean()(callback);
} else if (process.argv[2] === "--delete-empty") {
    delete_empty_directories()(callback);
} else {
    parseq.sequence([
        parseq.parallel_object({
            transformer: load_transformer(process.argv[2]),
            file_paths: list_tracked_files()
        }),
        function transform_all_requestor(callback, {transformer, file_paths}) {
            const tasks = file_paths.map(function (file_path) {
                return {
                    file_path,
                    requestor(callback) {
                        return transformer()(callback, {
                            file_path,
                            read_source_requestor: read_source(),
                            write_source_requestor: write_source(file_path)
                        });
                    }
                };
            });

            return parseq.sequence([
                clean(),
                parseq.parallel(tasks.map((task) => task.requestor)),
                function rename_all_requestor(callback, file_paths) {
                    return parseq.parallel(
                        file_paths.map(function (to, index) {
                            return rename_file(tasks[index].file_path, to);
                        })
                    )(callback);
                },
                stage_all()
            ])(callback);
        }
    ])(callback);
}
```

**Comment by kaizhu256** [comment]
here's a solution that passes jslint (but probably not in spirit ;)

1. create a separate script `dynamic-import.js`, who only purpose is to dynamically import esmodules:

```javascript
/*
 * dynamic-import.js
 */
/*jslint browser*/
/*global import, module*/
module.exports = function (pathname) {
/*
 * this function will dynamically-import <pathname>
 */
    return import(pathname);
};
```

2. then use above script to dynamically-import your desired module in main-script:

```diff
diff --git a/main.mjs b/main.mjs
--- a/main.mjs
+++ b/main.mjs
@@ -2,6 +2,11 @@
 import parseq from "./parseq.js";
 import make_capabilities from "./capabilities.js";

+// jslint-hack to get around "Undeclared 'import'"
+import {createRequire} from "module";
+let require = createRequire(process.cwd() + "/main.mjs");
+let dynamic_import = require("./dynamic-import.js");
+
 // Runs a transform over all files tracked by git in the current working
 // directory.
 // Usage: $0 <transformer_name>
@@ -35,7 +40,7 @@ import make_capabilities from "./capabilities.js";

 function load_transformer(name) {
     return function load_transformer_requestor(callback) {
-        return import(
+        return dynamic_import(
             "./transformers/" + name.replace("/", "") + ".js"
         ).then(
             (module) => callback(module.default)
```

3. here's the full-script `main.js`:

```javascript
/*
 * main.mjs
 */

/*jslint node */
import parseq from "./parseq.js";
import make_capabilities from "./capabilities.js";

// jslint-hack to get around "Undeclared 'import'"
import {createRequire} from "module";
let require = createRequire(process.cwd() + "/main.mjs");
let dynamic_import = require("./dynamic-import.js");

// Runs a transform over all files tracked by git in the current working
// directory.
// Usage: $0 <transformer_name>
// where <transformer_name> selects to a module exporting a requestor factory
// like the example below.
//
// Contrived example: uppercase the source code of all JS files, and add a
// ".upper" file extension.
//
// function upper_case_transformer() {
//     return function upper_case_transformer_requestor(
//         callback,
//         {file_path, read_source_requestor, write_source_requestor}
//     ) {
//         if (!file_path.endsWith(".js")) {
//             return callback(file_path);
//         }

//         return parseq.sequence([
//             read_source_requestor,
//             function transform_source_requestor(callback, source) {
//                 return callback(source.toUpperCase());
//             },
//             write_source_requestor,
//             function transform_filename_requestor(callback) {
//                 return callback(file_path + ".upper");
//             }
//         ])(callback, file_path);
//     };
// }

function load_transformer(name) {
    return function load_transformer_requestor(callback) {
        return dynamic_import(
            "./transformers/" + name.replace("/", "") + ".js"
        ).then(
            (module) => callback(module.default)
        ).catch(
            (err) => callback(undefined, err)
        );
    };
}

const {
    clean,
    delete_empty_directories,
    list_tracked_files,
    read_source,
    write_source,
    rename_file,
    stage_all
} = make_capabilities(process.cwd());

function callback(value, reason) {
    if (value === undefined) {
        console.error(reason);
        return process.exit(1);
    }

    return process.exit(0);
}

if (process.argv[2] === "--clean") {
    clean()(callback);
} else if (process.argv[2] === "--delete-empty") {
    delete_empty_directories()(callback);
} else {
    parseq.sequence([
        parseq.parallel_object({
            transformer: load_transformer(process.argv[2]),
            file_paths: list_tracked_files(),
            clean: clean()
        }),
        function transform_source_requestor(
            callback,
            {transformer, file_paths}
        ) {
            parseq.parallel(
                file_paths.map(function (file_path) {
                    return function requestor(callback) {
                        return transformer()(callback, {
                            file_path,
                            read_source_requestor: read_source(),
                            write_source_requestor: write_source(file_path)
                        });
                    };
                })
            )(function (requested_paths, reason) {
                if (requested_paths === undefined) {
                    return callback(undefined, reason);
                }

                return callback(
                    file_paths.map(function (file_path, index) {
                        return {
                            from: file_path,
                            to: requested_paths[index]
                        };
                    }).filter(function ({from, to}) {
                        return from !== to;
                    })
                );
            });
        },
        function rename_all_requestor(callback, renames) {
            return parseq.parallel(
                renames.map(function ({from, to}) {
                    return rename_file(from, to);
                })
            )(callback);
        },
        stage_all()
    ])(callback);
}
```

4. run it in commandline `node --experimental-modules main.mjs my-dynamic-module.mjs`

**Comment by diachedelic** [comment]
@kaizhu256 a creative solution! However, I am only interested in passing JSLint *in* spirit.

**Comment by diachedelic** [comment]
I realised that my program was structured incorrectly. I have changed it to use multiple entry points, which each statically import the shared code above – thereby avoiding the use of `import()` entirely.

**Comment by douglascrockford** [comment]
Do you feel that JSLint helped you to find a better solution, or did it just frustrate you?

**Comment by diachedelic** [comment]
Yes, it helped me improve my program in several ways, but the *Undeclared 'import'.* message was misleading. What I really needed to be told was that `import()` must only take a string literal.  I think parsing `import()` as a statement leads to confusing behaviour:

```js
// Minimal usage is ok
import("./grandmaster.js").then(start_game);
// JSLint: ok

// But error handling is not allowed
import("./grandmaster.js").then(start_game).catch(function (ignore) {
    alert("Pick another opponent.");
});
// JSLint: Expected ';' and instead saw '.'.

// Promise chaining is not allowed
import("./grandmaster.js").then(
    function (module) {
        return module.default.prepare_to_play();
    }
).then(
    start_game
);
// JSLint: Expected ';' and instead saw '.'.

// Assignment is not allowed
const grandmaster_promise = import("./grandmaster.js");
// JSLint: Undeclared 'import'.

// But is trivially circumvented
const grandmaster_promise = new Promise(function (resolve) {
    import("./grandmaster.js").then(resolve);
});
// JSLint: ok

// Returning the promise is not allowed
function challenge_grandmaster() {
    return import("./grandmaster.js").then(start_game);
}
// JSLint: Undeclared 'import'.
```

Which leads me to wonder – why is `import()` allowed at all?
